torch/_functorch/aot_autograd.py

```diff
@@ -138,86 +138,365 @@ def posthook(grad_input, grad_output):
-    # It's optional here, for cases where the user directly returns an input as an output.
-    # If output_type == non_alias, then these fields are also always None.
-    tensor_meta: Optional[Tensor]
-
-# This class tells us about how to perform a metadata mutation on forward inputs.
-# it only applies to forward inputs that experience metadata-only mutations
+    base_idx: Optional[int]
+    # Only important when output_type == OutputType.alias_of_input
+    # Tells the runtime code to return args[base_idx] *exactly*, and not generate a view off of it.
+    is_input_tensor: bool
+    # Only important when output_type == OutputType.alias_of_intermediate
+    # Tells the code that traces the joint that we should append output._base as an output
+    # in the forward graph.
+    add_base_to_outputs: bool
+
+# This class tells us info about user inputs.
 @dataclass(frozen=True)
 class InputAliasInfo:
-    # This object gives us information about how to perform a metadata-mutation
-    # on original_fwd_inputs[base_idx]
-    #   (This is an index into the inputs *before* we make synthetic bases)
-    base_idx: int
-    # sizes, strides and storage offset of the aliased output are all returned as actual (sym)ints
-    # in the compiled forward. These indices tell us where in the forward outputs to grab them.
-    sizes_idx: int
-    strides_idx: int
-    storage_offset_idx: int
-    # We store the actual output alias that we traced in the forward (should be a fake tensor)
-    # to grab any other non-symbolic properties on the output alias, like requires_grad.
-    tensor_meta: Tensor
+    mutation_type: MutationType
 
 # This class encapsulates all aliasing + mutation info we need about the forward graph
 # See a more detailed overview of the edge case handling at
 # https://docs.google.com/document/d/19UoIh_SVrMy_b2Sx5ZaeOJttm6P0Qmyss2rdBuyfoic/edit
 @dataclass(frozen=True)
 class ViewAndMutationMeta:
-    # length: # user forward inputs
-    # For every input, tells us whether the input:
-    # (a) is not mutated
-    # (b) only metadata is mutated
-    # (c) data (and maybe metadta) is mutated
-    mutated_input_info: List[MutationType]
-    # length: (# inputs of the user forward)
-    # metadata_mutation_input_info[i] is not None <====> mutated_input_info[i] == MutationType.metadata_only
-    # We stash the updated FakeTensor that we traced with in the forward in here,
-    # that way we can use it to replay the metadata mutation
-    metadata_mutation_input_info: List[Optional[InputAliasInfo]]
-    # length: # outputs in the compiled forward (not including output alias symints). Equal to:
-    # length: (# inputs w data mutations) + (# outputs that don't alias inputs)
+    # length = # user inputs
+    # This gives us info about every input, and what sort of mutation happened to it (if any)
+    input_info: List[InputAliasInfo]
+    # length = # user outputs + # intermediates that are returned as graph outputs
```

Comment:
It probably better not to talk about "length" but to talk about the layout, since, e.g., I assume the order you are summing up the subcomponents is the order you access them here.

But, if there is not a reason why you cannot do this (maybe you are forced to flatten all the lists), it would be better to give these mini structures (maybe NamedTuple). So you would have, instead of List:

```
class OutputAliasInfos(NamedTuple):
    user_outputs: List[OutputAliasInfo]
    intermediates: List[OutputAliasInfo]
```

and that would make it immediately clear how things are grouped into various pieces.

Reply:
agreed - intermediates go at the end. In fact:

* we don't need to store `OutputAliasInfo` on the intermediates, since they are always of type "non_alias"
* The only other piece of info that this PR currently stores on graph intermediates is their `requires_grad` during tracing. But I think it might be true that we only need to return the intermediate in the first place if its require_grad field is False. I can think of one example where a tensor's ._base.requires_grad is False:

```
def f(x):
    ...
    leaf = torch.ones_like(x)
    out_view = leaf.view(-1)
    leaf.requires_grad_(False) #this. is only valid to do on leaf tensors.
    return out_view, ...
```

But in this case, it's safe for our compiled graph to directly return `out_view`

So maybe I can get away with not storing any metadata about the intermediate in the first palce.

Reply:
I ended up not adding the NamedTuple mini structure, because I just completely removed all metadata that I was storing about graph intermediates.

(Haven't pushed changes yet, will let you know when I'm ready for another round)